tensorflow/compiler/xla/service/algebraic_simplifier.cc

```diff
@@ -1665,6 +1668,228 @@ StatusOr<HloInstruction*> AlgebraicSimplifierVisitor::OptimizeDotOfGather(
+  // remainder of this function easier.
+  auto dnums = dot->dot_dimension_numbers();
+  auto lhs_contracting_dims = dnums.lhs_contracting_dimensions();
+  auto rhs_contracting_dims = dnums.rhs_contracting_dimensions();
+  auto* lhs = dot->mutable_operand(0);
+  auto* rhs = dot->mutable_operand(1);
+  if (dot->operand(0)->IsConstant()) {
+    std::swap(lhs, rhs);
+    std::swap(lhs_contracting_dims, rhs_contracting_dims);
+  }
+
+  // Require single contracting dim to make the implementation easier to
+  // track contracting dims.
+  if (dnums.lhs_contracting_dimensions_size() != 1) {
+    return nullptr;
+  }
+
+  // Pattern match Dot(reshape(transpose(input), constant))
+  HloInstruction* reshape;
+  HloInstruction* transpose;
+  HloInstruction* input;
+  HloInstruction* constant;
+  if (!Match(lhs,
+             m::Reshape(&reshape, m::Transpose(&transpose, m::Op(&input)))) ||
+      !Match(rhs, m::Constant(&constant))) {
+    return nullptr;
+  }
+
+  // Check if reshape squishes some dims into one dim, and that if this one
+  // dim is the dot's lhs contracting dim.
+  CHECK(lhs_contracting_dims.size() == 1);
+  const std::vector<std::pair<int64, int64>>& unmodified_dims =
+      ShapeUtil::DimensionsUnmodifiedByReshape(lhs->operand(0)->shape(),
+                                               lhs->shape());
+  if (unmodified_dims.size() != lhs->shape().rank() - 1) {
+    return nullptr;
+  }
+  for (const auto &pair : unmodified_dims) {
+    if (pair.second == lhs_contracting_dims[0]) {
+      // If the contracting dim is unmodified, and only one dim is modified,
+      // then that dim must be a non-contracting dim.
+      return nullptr;
+    }
+  }
+  // Virtually pull the reshape into the dot. Now the dot is equivalent to a
+  // new dot with "unsquished" lhs contracting dims. We don't need to actually
+  // create a new dot instruction. We can just keep track of lhs and
+  // lhs_contracting_dims.
+  CHECK(lhs->operand(0)->shape().rank() > lhs->shape().rank());
+  for (int64 i = 1; i <= lhs->operand(0)->shape().rank() - lhs->shape().rank();
+       ++i) {
+    lhs_contracting_dims.Add(lhs_contracting_dims[0] + i);
+  }
+  lhs = lhs->mutable_operand(0);
+
+  // Check if lhs_transpose only permutes the contracting dims.
+  const auto& transpose_dims = lhs->dimensions();
+  for (int64 i = 0; i < transpose_dims.size(); ++i) {
+    if (transpose_dims[i] != i &&
+        !(absl::c_linear_search(lhs_contracting_dims, i) &&
+          absl::c_linear_search(lhs_contracting_dims, transpose_dims[i]))) {
+      return nullptr;
+    }
+  }
+  // Virtually pull the transpose into the dot. Now the dot is equivalent to
+  // a new dot with new lhs transpose dims equal to the transpose dims within
+  // the old lhs contracting dims.
+  auto old_lhs_contracting_dims = lhs_contracting_dims;
+  lhs_contracting_dims.Clear();
+  for (const auto dim : old_lhs_contracting_dims) {
+    lhs_contracting_dims.Add(lhs->dimensions(dim));
+  }
```

Comment:
Can we use the Permute() function in util.h, or does that not work?

Reply:
Not sure I understand. Are you suggesting that we use Permute() to permute the contracting dims here, as oppose to the current way of copying transpose dims? I think it would take some effort to compute the permutation vector for the Permute, so it may be easier to just copy the transpose dims.

Reply:
> I think it would take some effort to compute the permutation vector for the Permute

The permutation vector is not `lhs->dimensions()`?

Reply:
For example, let's say we have lhs transpose (type omitted) A = [a, d, b, c] transpose([a, b, c, d] B) dims={0, 3, 1, 2}. The contracting dims of B is {1, 2, 3}. The current implementation here uses {1, 2, 3} to index into the transpose dims to get A's contracting dims {3, 1, 2}. Maybe I'm missing something, as I still don't see how to permute B's contracting dims to get A's contracting dims using lhs->dimensions().

Reply:
Given

```
contracting_dims=[1,2,3]
perm=[0,3,1,2]
result=[3,1,2]
```

The function here seems to be

```
result[i] = perm[contracting_dims[i]]
```

?

This looks pretty similar to the function `ComposePermutations` in util.h?  (Looks like I pointed at the wrong helper function.)  Right now ComposePermutations requires the two perms to be of equal length, but that restriction could be lifted (separate PR), maybe with a new function, or maybe the same function.

Reply:
I keep the implement here and add a comment. I do not use helper function here since it would need some change, and also the helper function expect a absl::Span<const int64> and it is not very friendly to RepeatedField<int64>.